File: DFS-BFS-4travelPath/travelPath.py
import random
import time


'''
airport_num이 500이면 오래 걸리고 
1000 이상이면 재귀 깊이 제한 오류 발생
제한 깊이 제한을 느슨하게 하거나 스택 자료 구조를 이용한 반복문 활용 가능
제일 느림..

dfs 활용
방문 처리 활용
dfs진행하면서 path와 방문 리스트를 업데이트함 (얕은 복사 주의)
전역 paths 2차원 배열 활용
'''

paths = []


def dfs(path, tickets, used_ticket):
    if False in used_ticket:          # 사용하지 않은 티겟이 있다면 함수 게속 진행
        # 다음 티겟 찾기
        for i, ticket in enumerate(tickets):
            next_path = path[:]                 # path값을 유지하기 위하여 next_path에 복사
            # used_tickek 값을 유지하기 위하여 next_used_ticket에 복사
            next_used_ticket = used_ticket[:]
            if path[-1] == ticket[0]:            # 다음 도착지를 가진 티겟을 발견하면
                if next_used_ticket[i] == False:   # 사용되지 않은 티겟임을 확인하고
                    # 다음 도착지를 next_path에 추가
                    next_path.append(ticket[1])
                    next_used_ticket[i] = True                # 그 티겟은 사용 처리함
                    # 경로와 티겟 사용여부 리스트를 업데이트하여 dfs에 넘김
                    dfs(next_path, tickets, next_used_ticket)
    else:            # 모든 티켓을 사용했다면  paths에 path를 추가하고 리턴
        paths.append(path)


def solution(tickets):
    global paths
    used_ticket = [False] * len(tickets)

    path = ["ICN"]  # path ICN 추가
    dfs(path, tickets, used_ticket)  # dfs 탐색 시작
    paths = sorted(paths)    # paths를 알파펫 순으로 정렬
    print(paths)

    return paths[0]   # 알파벳 순서가 가장 앞서는 경로를 리턴


airport_num = random.randint(3, 10000)  # 주어진 공항 수는 3개 이상 10,000개 이하
# airport_num = 500
tickets = []
fr = "ICN"
for _ in range(airport_num):
    to = ""
    for _ in range(3):
        to += chr(random.randint(65, 90))  # A(65) ~ Z(90)
    ticket = [fr, to]      # [출발, 도착] 티켓
    fr = to
    tickets.append(ticket)  # 티켓 리스트
random.shuffle(tickets)  # 티켓 리스트 셔플
# print(tickets)

tickets = [["ICN", "SFO"], ["ICN", "ATL"], [
    "SFO", "ATL"], ["ATL", "ICN"], ["ATL", "SFO"]]

start_time = time.time()
result = solution(tickets)
print("--- %s seconds ---" % (time.time() - start_time))
# print(result)

File: DFS-BFS-4-travelPath/travelPath.cpp
#include <bits/stdc++.h>

using namespace std;

/*
// 방문 처리
// 다음 노드들에 대해 방문 하지 않았다면 dfs 방문함

vector<bool> visited(n, 0); // 방문처리 초기화 // 1차원 배열 말고 vector 이용하면 디버깅할 때 값 보기 편함

알파벳 순으로 티겟 정렬

// 최종 경로가 완성되면 paths에 path 추가함

다음 노드들에 대해
    방문하지 않았다면
        dfs 탐색
(참조자로 visited 끌고가며 업데이트)
(path도 끌고가며 업데이트)

해당 dfs 재귀가 소멸하는 과정에서 반복문이 반복되며 다른 다음 노드로 진입할 수도 있고
반복이 끝났다면  visited[v] = false;  path.pop_back(); 업데이트하고  소멸됨
*/

vector<vector<string>> paths;

void dfs(vector<vector<string>> &tickets, vector<bool> &visited, int v, string fr, vector<string> &path)
{
    // 최종 경로가 완성되면
    if (path.size() == tickets.size() + 1)
    {
        paths.push_back(path);
        visited[v] = false;
        path.pop_back();
        return;
    }

    // 다음 노드들에 대해
    for (int i = 0; i < tickets.size(); i++)
    {
        if (fr == tickets[i][0])
        {
            // 방문하지 않았다면
            if (!visited[i])
            {
                visited[i] = true;
                path.push_back(tickets[i][1]);
                dfs(tickets, visited, i, tickets[i][1], path);
            }
        }
    }
    // 방문할 다음 노드가 없으므로
    visited[v] = false;
    path.pop_back();
}

vector<string> solution(vector<vector<string>> tickets)
{
    int n = tickets.size();
    vector<bool> visited(n, 0); // 방문처리 초기화 // 1차원 배열 말고 vector 이용하면 디버깅할 때 값 보기 편함

    sort(tickets.begin(), tickets.end()); // 알파벳 순으로 티겟 정렬

    vector<string> path;
    path.push_back("ICN");
    dfs(tickets, visited, 0, "ICN", path);

    return paths[0];
}

int main(void)
{
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<int> dis_airport_num(3, 10000); // 주어진 공항 수는 3개 이상 10,000개 이하
    uniform_int_distribution<int> dis_alphabet(65, 90);      // 'A' ~ 'Z'

    int airport_num = dis_airport_num(gen);
    airport_num = 800;

    string fr = "ICN";
    vector<vector<string>> tickets;
    for (int i = 0; i < airport_num; i++)
    {
        vector<string> ticket;
        string to = "";
        ticket.push_back(fr);
        for (int j = 0; j < 3; j++)
        {
            to += char(dis_alphabet(gen));
        }
        if (fr != to) // 티켓의 출발지와 도착지가 달라지도록 처리
        {
            ticket.push_back(to);
            fr = to;
            tickets.push_back(ticket);
        }
        else
            i--;
    }

    shuffle(tickets.begin(), tickets.end(), gen); // 배열의 순서를 랜덤으로 셔플

    // tickets = {{"ICN", "SFO"}, {"ICN", "ATL"}, {"SFO", "ATL"}, {"ATL", "ICN"}, {"ATL", "SFO"}};

    // cout << "tickets: ";
    // for (int i = 0; i < tickets.size(); i++)
    // {
    //     cout << "{" << tickets[i][0] << ", " << tickets[i][1] << "}" << ' ';
    // }
    // cout << endl;

    vector<string> result;
    auto start_time = chrono::steady_clock::now();

    result = solution(tickets);

    auto end_time = chrono::steady_clock::now();
    chrono::duration<double> elapsed_seconds = end_time - start_time;
    cout << "--- " << elapsed_seconds.count() << " seconds ---" << endl;
    cout << "result: ";
    for (int i = 0; i < result.size(); i++)
    {
        cout << result[i] << " ";
    }
    cout << endl;

    return 0;
}